src/mame/brother/lw30.cpp

```diff
@@ -0,0 +1,1105 @@
+		// bit 5: cleared in rst28_06; motor-on?
+		LOGFLOPPY("%s: read %02X from IO 88\n", machine().describe_context(), io_88);
+		return io_88;
+	}
+	void io_88_w(uint8_t data) {
+		LOGFLOPPY("%s: write %02X to IO 88\n", machine().describe_context(), data);
+		io_88 = data;
+		floppy->get_device()->mon_w((io_88 & (1 << 5)) == 0);
+	}
+
+	uint8_t floppy_status_r() { // 90
+		// bit 7 set; data ready from floppy
+		// bit 6 clear; unknown meaning
+		// bit 5 clear; unknown meaning
+		// bit 4 clear; unknown meaning
+		// bit 3-0: stepper motor
+		LOGFLOPPY("%s: read %02X from IO 90\n", machine().describe_context(), floppy_control);
+		return floppy_control;
+	}
+	void floppy_stepper_w(uint8_t data) {
+		LOGFLOPPY("%s: write %02X to IO 90\n", machine().describe_context(), data);
+		// write directly to 4-wire bipolar stepper motor (see stepper_table)
+		// a rotation to the left means decrease quarter-track
+		// a rotation to the right means increase quarter-track
+		auto rol4 = [](uint8_t d) { return ((d << 1) & 0b1111) | ((d >> 3) & 0b0001); };
+		auto ror4 = [](uint8_t d) { return ((d >> 1) & 0b0111) | ((d << 3) & 0b1000); };
+		auto old_track = floppy_steps / 4;
+		switch(data & 0xf) {
+		case 0b0011:
+		case 0b0110:
+		case 0b1100:
+		case 0b1001:
+			if((data & 0x0f) == rol4(floppy_control))
+				floppy_steps--;
+			else if((data & 0x0f) == ror4(floppy_control))
+				floppy_steps++;
+			else
+				LOGFLOPPY("%s: illegal step %02x=>%02x\n", machine().describe_context(), floppy_control, data);
+			break;
+		default:
+			LOGFLOPPY("%s: initial step %02x=>%02x\n", machine().describe_context(), floppy_control, data);
+			break;
+		}
+		auto new_track = floppy_steps / 4;
+		auto floppy_device = floppy->get_device();
+		if(new_track != old_track) {
+			floppy_device->dir_w(new_track < old_track);
+			floppy_device->stp_w(true);
+			floppy_device->stp_w(false);
+		}
+		LOGFLOPPY("%s: floppy_steps=%3d => old_track=%2d new_track=%2d cyl=%2d\n", machine().describe_context(), floppy_steps, old_track, new_track, floppy_device->get_cyl());
+		assert(floppy_device->get_cyl() == new_track);
+		floppy_control = (floppy_control & 0xf0) | (data & 0x0f);
+	}
+
+	uint8_t io_98_r() {
+		// mirrored in RAM
+		// bit 0: cleared in rst28_06 in mirror
+		// bit 2: cleared before formatting in mirror; set after formatting
+		// bit 3: cleared before formatting in mirror
+		// bit 4: cleared before writing in mirror; set after writing
+		if(io_88 & 0b10)
+			floppy_read_until_zerobit = true;
+		else
+			floppy_read_until_zerobit = false;
+
+		LOGFLOPPY("%s: read %02X from IO 98\n", machine().describe_context(), io_98);
+		return io_98;
+	}
+	void io_98_w(uint8_t data) {
+		LOGFLOPPY("%s: write %02X to IO 98\n", machine().describe_context(), data);
+		io_98 = data;
+	}
+
+	uint8_t illegal_io_r(offs_t offset, uint8_t mem_mask = ~0) {
+		LOGFLOPPY("%s: unmapped IO read from %0*X & %0*X\n", machine().describe_context(), 4, offset + 0x40, 2, mem_mask);
+		return 0;
+	}
```

Comment:
It’s probably worth guarding the `LOGFLOPPY` with `if (!machine().side_effects_disabled()` to avoid log spam when a debugger memory window is open.